### DISCUSSION TITLE
Global fetch store

### DISCUSSION THREAD

**Discussion by koronowicz** [withastro/roadmap#603] [Proposal]
### Body

# Summary

Ability to fetch data before generating pages, then refer to it in page, components.

# Background & Motivation

When working in bigger project, it's easier to fetch data before generation. 

# Goals

It'll eliminate a lot of multiple similar requests to get data that was fetch before.

**Comment by theotterlord** [comment]
Does [middleware](https://docs.astro.build/en/guides/middleware/) help alleviate this? You can make your requests in the first part of your middleware function, then save the results to the `locals` context. This makes that data available via `Astro.locals` in Astro components, or `context.locals` in API endpoints.

```ts
export function onRequest ({ locals, request }, next) {
  // intercept response data from a request
  // fetch shared contextual data by modifying `locals`
  locals.data = await fetch('https://example.com/data.json').then(r => r.json())

  // render the result by calling `next()`
  return next()
};
```

**Reply by koronowicz** [reply]
Thanks, that's working great. It's also good to use it for development with node-fetch-cache.

**Reply by regisphilibert** [reply]
I'm curious if you can make the Astro.locals persistant during build and available for every page. This looks like Astro.locals will be shared across components generated for one page, but not all.

So say I want to fetch the menu items form the API which will be printed on every page, is _middleware_ the solution to avoid calling the API for every page?

**Reply by ematipico** [reply]
> I'm curious if you can make the Astro.locals persistant during build and available for every page. This looks like Astro.locals will be shared across components generated for one page, but not all.
> 
> So say I want to fetch the menu items form the API which will be printed on every page, is _middleware_ the solution to avoid calling the API for every page?

No, Astro.locals was designed to persist only for the time of a request and shouldn't be used for a long-running data store.

You can still use the middleware module to have a hoisted object where you can save your stuff, although it won't probably work in a serverless environment.

If you're interested in a long-running data store, check out https://github.com/withastro/roadmap/discussions/670
